src/geometry.F90

```diff
@@ -759,34 +648,47 @@ end subroutine cross_lattice
 
     type(Particle), intent(inout) :: p
     real(8),        intent(out)   :: dist
     integer,        intent(out)   :: surface_crossed
-    integer,        intent(out)   :: lattice_crossed
-
-    integer :: i            ! index for surface in cell
-    integer :: index_surf   ! index in surfaces array (with sign)
-    real(8) :: x,y,z        ! particle coordinates
-    real(8) :: u,v,w        ! particle directions
-    real(8) :: d            ! evaluated distance
-    real(8) :: x0,y0,z0     ! coefficients for surface
-    real(8) :: r            ! radius for quadratic surfaces
-    real(8) :: tmp          ! dot product of surface normal with direction
-    real(8) :: a,b,c,k      ! quadratic equation coefficients
-    real(8) :: quad         ! discriminant of quadratic equation
-    logical :: on_surface   ! is particle on surface?
+    integer,        intent(out)   :: lattice_translation(3)
+
+    integer :: i                  ! index for surface in cell
+    integer :: index_surf         ! index in surfaces array (with sign)
+    integer :: i_xyz(3)           ! lattice indices
+    integer :: level_surf_cross   ! surface crossed on current level
+    integer :: level_lat_trans(3) ! lattice translation on current level
+    real(8) :: x,y,z              ! particle coordinates
+    real(8) :: xyz_t(3)           ! local particle coordinates
+    real(8) :: beta, gama         ! skewed particle coordiantes
+    real(8) :: u,v,w              ! particle directions
+    real(8) :: beta_dir           ! skewed particle direction
+    real(8) :: gama_dir           ! skewed particle direction
+    real(8) :: edge               ! distance to oncoming edge
+    real(8) :: d                  ! evaluated distance
+    real(8) :: d_lat              ! distance to lattice boundary
+    real(8) :: d_surf             ! distance to surface
+    real(8) :: x0,y0,z0           ! coefficients for surface
+    real(8) :: r                  ! radius for quadratic surfaces
+    real(8) :: tmp                ! dot product of surface normal with direction
+    real(8) :: a,b,c,k            ! quadratic equation coefficients
+    real(8) :: quad               ! discriminant of quadratic equation
+    logical :: on_surface         ! is particle on surface?
     type(Cell),       pointer, save :: cl => null()
     type(Surface),    pointer, save :: surf => null()
-    type(Lattice),    pointer, save :: lat => null()
+    class(Lattice),   pointer, save :: lat => null()
     type(LocalCoord), pointer, save :: coord => null()
     type(LocalCoord), pointer, save :: final_coord => null()
-!$omp threadprivate(cl, surf, lat, coord, final_coord)
+    type(LocalCoord), pointer, save :: parent_coord => null()
+!$omp threadprivate(cl, surf, lat, coord, final_coord, parent_coord)
 
     ! inialize distance to infinity (huge)
     dist = INFINITY
-    lattice_crossed = NONE
+    d_lat = INFINITY
+    d_surf = INFINITY
+    lattice_translation = (/0, 0, 0/)
```

Comment:
I prefer bracket notation for arrays, i.e. `[0, 0, 0]`, which is equivalent to `(/0, 0, 0,/)` per the Fortran standard. Two fewer characters, and it's reminiscent of Python :) But the choice is yours.